src/Uno.UI/UI/Xaml/Controls/Image/ImageSizeHelper.cs

```diff
@@ -1,122 +1,159 @@
 					break;
 				case HorizontalAlignment.Stretch:
-					child.X = isForcedLeft ? 0 : (parent.Width * 0.5f) - (child.Width * 0.5f);
+					child.X = GetPositionForDimension(isForcedLeft, -1, (finalSize.Width - child.Width) * 0.5f);
 					break;
 			}
 
 			// In order to match UWP behaviors, in some specific cases the image must be top align
-			var isForcedTop = (image.Stretch == None || image.Stretch == UniformToFill) && parent.Height <= child.Height;
+			//var isForcedTop = (stretch == None || stretch == UniformToFill) && finalSize.Height <= child.Height;
+			var isForcedTop = false;
 
 			switch (image.VerticalAlignment)
 			{
 				case VerticalAlignment.Top:
 					child.Y = 0;
 					break;
 				case VerticalAlignment.Bottom:
-					child.Y = isForcedTop && !Double.IsNaN(image.Height) ? 0 : parent.Height - child.Height;
+					child.Y = GetPositionForDimension(isForcedTop, image.Height, finalSize.Height - child.Height);
 					break;
 				case VerticalAlignment.Center:
-					child.Y = isForcedTop && !Double.IsNaN(image.Height) ? 0 : (parent.Height * 0.5f) - (child.Height * 0.5f);
+					child.Y = GetPositionForDimension(isForcedTop, image.Height, (finalSize.Height - child.Height) * 0.5f);
 					break;
 				case VerticalAlignment.Stretch:
-					child.Y = isForcedTop ? 0 : (parent.Height * 0.5f) - (child.Height * 0.5f);
+					child.Y = GetPositionForDimension(isForcedTop, -1, (finalSize.Height - child.Height) * 0.5f);
 					break;
 			}
 		}
 
 		public static (double x, double y) BuildScale(this Image image, Size destinationSize, Size sourceSize)
 		{
-			if (image.Stretch != Stretch.None)
+			return BuildScale(image.Stretch, destinationSize, sourceSize);
+		}
+
+		internal static (double x, double y) BuildScale(Stretch stretch, Size destinationSize, Size sourceSize)
+		{
+			if (stretch != None)
 			{
 				var scale = (
 					x: destinationSize.Width / sourceSize.Width,
 					y: destinationSize.Height / sourceSize.Height
 				);
 
-				switch (image.Stretch)
+				if (double.IsInfinity(scale.x))
+				{
+					if (double.IsInfinity(scale.y))
+					{
+						return (1.0d, 1.0d);
+					}
+
+					scale.x = scale.y;
+				}
+				else if (double.IsInfinity(scale.y))
+				{
+					scale.y = scale.x;
+				}
+
+				switch (stretch)
 				{
-					case Stretch.UniformToFill:
+					case UniformToFill:
 						var max = Math.Max(scale.x, scale.y);
 						scale = (max, max);
 						break;
 
-					case Stretch.Uniform:
+					case Uniform:
 						var min = Math.Min(scale.x, scale.y);
 						scale = (min, min);
 						break;
 				}
 
-				return (
-					double.IsNaN(scale.x) || double.IsInfinity(scale.x) ? 1 : scale.x
-					, double.IsNaN(scale.y) || double.IsInfinity(scale.y) ? 1 : scale.y
-				);
+				var scaleX = double.IsNaN(scale.x) || double.IsInfinity(scale.x) ? 1.0d : scale.x;
+				var scaleY = double.IsNaN(scale.y) || double.IsInfinity(scale.y) ? 1.0d : scale.y;
```

Comment:
I think that at this point we cannot have an infinity anymore.

Reply:
Fixed